### PR TITLE
wasi: Make WasiCtxBuilder::socket_addr_check closure async

### LINKED ISSUE #8714: Make wasmtime_wasi's `SocketAddrCheck` async
Currently, the [`WasiCtxBuilder::socket_addr_check`](https://docs.rs/wasmtime-wasi/21.0.1/wasmtime_wasi/struct.WasiCtxBuilder.html#method.socket_addr_check) callback is synchronous. While this is often reasonable, I'm working on some code that wants to resolve inputs to this check asynchronously, which requires either delaying construction of the `WasiCtx` (which delays execution of the guest) or blocking the thread within the check.

Could this allow async callbacks? Either updating the existing signature (presumably breaking existing code, but with a trivial fix) or adding a new e.g. `socket_addr_async_check` method?

cc @rylev

### PR DESCRIPTION
**Pull request by lann**
Fixes #8714

### CODE REVIEW

**Review by pchickey — APPROVED** [review]

**Comment by lann** [comment]
This still needs the sync impls, which I did not understand before but now do (hopefully).

**Comment by lann** [comment]
This seems code-complete, but I'm getting bindgen errors that I'm not quite understanding:

```
error[E0277]: the trait bound `U: async_io::wasi::sockets::network::Host` is not satisfied
  --> crates/wasi/src/bindings.rs:9:9
   |
9  | /         wasmtime::component::bindgen!({
10 | |             path: "wit",
11 | |             world: "wasi:cli/command",
12 | |             tracing: true,
...  |
41 | |             },
42 | |         });
   | |__________^ the trait `async_io::wasi::sockets::network::Host` is not implemented for `U`
```

File: crates/wasi/src/ctx.rs [hunk 1]
```diff
@@ -417,7 +418,10 @@ impl WasiCtxBuilder {
     /// while returning `false` will reject the connection.
     pub fn socket_addr_check<F>(&mut self, check: F) -> &mut Self
     where
-        F: Fn(&SocketAddr, SocketAddrUse) -> bool + Send + Sync + 'static,
+        F: Fn(SocketAddr, SocketAddrUse) -> Pin<Box<dyn Future<Output = bool> + Send + Sync>>
```
Comment (lann) [inline crates/wasi/src/ctx.rs:421]: Since we're breaking this signature anyway, I also `&SocketAddr` -> `SocketAddr` (which is `Copy`).

**Comment by lann** [comment]
@alexcrichton is working on bindgen improvements that ought to fix this.

**Comment by alexcrichton** [comment]
With https://github.com/bytecodealliance/wasmtime/pull/8721 I think this PR should be good to go

**Comment by lann** [comment]
Looks like this is ready to go now

**Comment by lann** [comment]
@alexcrichton OK to backport this to 22?

**Pull request merged**